mem/buffer_slice.go

```diff
@@ -224,3 +228,47 @@ func (w *writer) Write(p []byte) (n int, err error) {
 func NewWriter(buffers *BufferSlice, pool BufferPool) io.Writer {
 	return &writer{buffers: buffers, pool: pool}
 }
+
+// ReadAll reads from r until an error or EOF and returns the data it read.
+// A successful call returns err == nil, not err == EOF. Because ReadAll is
+// defined to read from src until EOF, it does not treat an EOF from Read
+// as an error to be reported.
+// Make sure to free the returned buffers even if you get an error from this function.
+func ReadAll(r io.Reader, pool BufferPool) (BufferSlice, error) {
+	var result BufferSlice
+	wt, ok := r.(io.WriterTo)
+	if ok {
+		// This is more optimal since wt knows the size of chunks it wants to write and, hence, we can allocate
+		// buffers of an optimal size to fit them. E.g. might be a single big chunk, and we wouldn't chop it into pieces.
+		w := NewWriter(&result, pool)
+		_, err := wt.WriteTo(w)
+		return result, err
+	}
+	for {
+		buf := pool.Get(readAllBufSize)
+		// We asked for 32KiB but may have been given a bigger buffer. Use all of it if that's the case.
+		*buf = (*buf)[:cap(*buf)]
```

Comment:
IIRC 32kb is one of the default sizes in the pool, specifically because io.ReadAll

Reply:
Yes, it's just a tiny tweak to make it work better for any passed pool.